Dynamic Networks: Document new API for hotplug/unplug

ormergi opened the pull request:
Update the dynamic network section on how to hotplug/unplug interfaces following the changes in Kubevirt.

AlonaKaplan commented on docs/operations/hotplug_interfaces.md:
````diff
@@ -79,27 +79,86 @@ Please refer to the
-```bash
-virtctl addinterface vm-fedora --network-attachment-definition-name new-fancy-net --name dyniface1
+provisioned, the user can request the interface hotplug operation by 
+editing the VM spec template and add the desired interface and network:
````
s/add/adding

ormergi replied:
Done

AlonaKaplan commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -79,27 +79,86 @@ Please refer to the
 
-**NOTE**: You can use the `--help` parameter for more information on each
-parameter.
+The interface and network will be removed from the corresponding VMI object as well by Kubevirt controller.
```
removed?

ormergi replied:
Thanks, fixed.

AlonaKaplan commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -79,27 +79,86 @@ Please refer to the
-parameter.
+The interface and network will be removed from the corresponding VMI object as well by Kubevirt controller.
+
+Alternatively, use JSON patch to automating the process: 
```
I don't think the json patch example is required. It is just another way to update a spec.

AlonaKaplan commented on docs/operations/hotplug_interfaces.md:
````diff
@@ -79,27 +79,86 @@ Please refer to the
+Alternatively, use JSON patch to automate the process:
 ```bash
-virtctl removeinterface vm-fedora --name dyniface1
+kubectl patch vm vm-fedora --type=json -p='[{"op": "replace", "path": "/spec/template/spec/domain/devices/interfaces/1/state", "value": "absent"}]'
````
Again I think the patch example is redundant.

AlonaKaplan commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -79,27 +79,86 @@ Please refer to the
-The subject interface state will be set as absent in the VM spec template and the running VMI object.  
 
 >**Note**: Existing VMs from version v0.59.0 and below does not support hot-unplug interfaces.
 
```
Migration based hotplug/unplug-
This section has to be updated as well. It has `virtctl` examples.

ormergi replied:
Done

AlonaKaplan commented:
Summarizing some general comment -
1. [ Hotplug is only supported for interfaces using the
`virtio` model connected through bridge binding](https://github.com/kubevirt/user-guide/pull/729/commits/a00b153a1f7f011a25ca885cd31e32c4437935ff#diff-1825ea10641a34a2e6f9d4b2eb70a37462af20304431775e6cc6ae8176082eddR7) is wrong. We support SRIOV hotplug as well.
2. We don't have docs for SRIOV hotplug/unplug. I don't think we need a separate section for it. Just update the existing to mention SRIOV hotplug (but not unplug) is also supported. Or adding a separate section mentioning SRIOV hotplug is supported and done exactly in the same way as bridge hot plug.

ormergi commented:
> Summarizing some general comment -
> 
>     1. [ Hotplug is only supported for interfaces using the
>        `virtio` model connected through bridge binding](https://github.com/kubevirt/user-guide/pull/729/commits/a00b153a1f7f011a25ca885cd31e32c4437935ff#diff-1825ea10641a34a2e6f9d4b2eb70a37462af20304431775e6cc6ae8176082eddR7) is wrong. We support SRIOV hotplug as well.
> 
>     2. We don't have docs for SRIOV hotplug/unplug. I don't think we need a separate section for it. Just update the existing to mention SRIOV hotplug (but not unplug) is also supported. Or adding a separate section mentioning SRIOV hotplug is supported and done exactly in the same way as bridge hot plug.

I updated the summary about hotplug and unplug support for bridge and SR-IOV interfaces 
and added small section about SR-IOV interfaces under hotplug section.

AlonaKaplan commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -2,10 +2,14 @@
-Machine (VM). Hotplug is only supported for interfaces using the
-`virtio` model connected through
-[bridge binding](http://kubevirt.io/api-reference/main/definitions.html#_v1_interfacebridge).
+KubeVirt supports hotplugging and unplugging network interfaces into to running Virtual Machine (VM). 
```
into a

ormergi replied:
Fixed

AlonaKaplan commented on docs/operations/hotplug_interfaces.md:
````diff
@@ -113,6 +117,40 @@ You can now check the VMI status for the presence of this new interface:
 kubectl get vmi vm-fedora -ojsonpath="{ @.status.interfaces }"
 ```
 
+### SR-IOV interfaces 
````
I would move the SRIOV hotplug to the migration based section and also mention the SRIOV hotplug is always migration based. Even if there is multus v4.

ormergi replied:
Done

AlonaKaplan commented:
/approve

Thanks!

EdDev commented on docs/operations/hotplug_interfaces.md:
````diff
@@ -79,27 +83,70 @@ Please refer to the
-provisioned, the user can request the interface hotplug operation. Please refer
-to the following snippet:
-```bash
-virtctl addinterface vm-fedora --network-attachment-definition-name new-fancy-net --name dyniface1
````
It may be good to add a note that `virtctl` commands used in v1.0 are dropped.
Not completely sure if it is a good move though.

ormergi replied:
It seems virctl hotplug/unplug commands are available in v1.0.0 and removed in v1.1.0
https://github.com/kubevirt/kubevirt/tree/v1.0.0/pkg/virtctl/network

EdDev commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -79,27 +83,70 @@ Please refer to the
 
-**NOTE**: You can use the `--help` parameter for more information on each
-parameter.
+The interface and network will be added to the corresponding VMI object as well by Kubevirt controller.
```
I would drop the `controller`, as it has many.
Alternatively you could mention `VM controller`, but entering into so much details may not be desired.

EdDev commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -79,27 +83,70 @@ Please refer to the
+    networks:
+      - name: defaultnetwork
+        pod: {}
+      - name: fancy-net
```
same

EdDev commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -79,27 +83,70 @@ Please refer to the
+        pod: {}
+      - name: fancy-net
+        multus:
+          networkName: new-fancy-net
```
same

EdDev commented on docs/operations/hotplug_interfaces.md:
````diff
@@ -79,27 +83,70 @@ Please refer to the
+          networkName: new-fancy-net
 ```
-The subject interface state will be set as absent in the VM spec template and the running VMI object.  
+The interface in the corresponding VMI object will be set with state 'absent' as well by Kubevirt controller.
````
same

EdDev commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -109,10 +156,33 @@ In case your cluster doesn't run Multus as [thick plugin](https://github.com/k8s
+    - name: defaultnetwork
+      pod: {}
+      # new network
+    - name: fancy-net
```
same

EdDev commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -109,10 +156,33 @@ In case your cluster doesn't run Multus as [thick plugin](https://github.com/k8s
+      # new network
+    - name: fancy-net
+      multus:
+        networkName: new-fancy-net
```
same

EdDev commented on docs/operations/hotplug_interfaces.md:
````diff
@@ -109,10 +156,33 @@ In case your cluster doesn't run Multus as [thick plugin](https://github.com/k8s
 ```
-At this point the new interface is added to the spec but will not be attached to the running VM. 
+
+At this point the interface and network will be added to the corresponding VMI object as well, but will not attached to the guest.
````
.. not **be** attached ...

EdDev commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -132,10 +202,67 @@ Once the migration is completed the VM will have the new interface attached.
 >**Note**: It is recommended to avoid performing migrations in parallel to a hotplug operation. 
 > It is safer to assure hotplug succeeded or at least reached the VMI specification before issuing a migration.
 
+#### SR-IOV interfaces
```
This appears under the `Migrate the VM` section, it looks wrong.

EdDev commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -132,10 +202,67 @@ Once the migration is completed the VM will have the new interface attached.
+    networks:
+      - name: defaultnetwork
+        pod: {}
+      - name: fancy-net
```
same

EdDev commented on docs/operations/hotplug_interfaces.md:
```diff
@@ -132,10 +202,67 @@ Once the migration is completed the VM will have the new interface attached.
+        pod: {}
+      - name: fancy-net
+        multus:
+          networkName: new-fancy-net
```
same

EdDev reviewed (approved):
Thanks!
